### ISSUE TITLE
Roslyn rule succeeded when it should have failed

### ISSUE THREAD

**Issue by zachhats** [dotnet/roslyn#79824]
(not sure if there is a more appropriate internal support queue, please let me know if there is)

**Version Used**: 
MSBuild 17.4

**Steps to Reproduce**:

We had a PR get in that enabled a Roslyn analyzer rule, but subsequently all runs started failing because there was a latent warning that was not flagged. The repro does not happen locally (I checked out the commit enabling the rule), it happened in CloudBuild. I talked to the CloudBuild team and they directed me to reach out to the Roslyn compiler team. Per David from CB team, the calls to csc are identical.

This is specifically about CA1868.

**Expected Behavior**:
Build failed

**Actual Behavior**:
Build succeeded

**Comment by zachhats** [comment]
@333fred please feel free to reach out to me directly - my alias is zahala. i can share the initial discussion with the CloudBuild team

**Comment by 333fred** [comment]
@zachhats can you get `complog`s for both the failing and succeeding cases? https://github.com/jaredpar/complog?tab=readme-ov-file#creating-compiler-logs

This will make it very easy to reproduce locally.

**Comment by zachhats** [comment]
@333fred Is it possible to create complogs from an already completed CloudBuild? I don't have access to the machine and the tool mentions:

```
They must be created on the same machine where the binary log was created but after creation they can be freely copied between machines.
```

If not, maybe I can set up complogs in case we run into anything similar going forward

**Comment by 333fred** [comment]
No, it must be created from the machine. Complogs capture the entire csc call and all inputs, so if there are no inputs left, it can't capture them. It would definitely be good to save them in the future; would it be possible to rerun the cloudbuild that failed and get a complog from it?

**Comment by KirillOsenkov** [comment]
We don't have a repro for this and we decided to drop this.

We think it might be a rare race condition in the analyzer:
https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs

But that's just the only explanation we can think of to explain the one instance where the warnings were expected, but not produced.

Feel free to close as not repro, because we're out of ideas. Adding a complog will just capture the correct behavior, and we can debug that locally if we want, but we don't have the complog for that one build where the warnings weren't produced.

**Comment by 333fred** [comment]
So, to be clear @KirillOsenkov: let's call the build that did not produce warnings A. If you rerun A in cloudbuild, exactly as it was, you _do_ get a warning?

**Comment by KirillOsenkov** [comment]
we do, and we also do get the warning locally, reliably

**Comment by KirillOsenkov** [comment]
hmm, I misunderstood the question. This is one thing we haven't tried. Let us try that.

**Comment by 333fred** [comment]
If that does pass (ie, does not produce a warning), then we should be able to troubleshoot if you can modify that build just enough to produce a complog.

**Comment by KirillOsenkov** [comment]
We reran that exact commit and it produced the warning now :( So definitely a fluke.

**Comment by zachhats** [comment]
we should theoretically be able to run the exact commit without caching, but we wouldn't be able to get complogs. talked to Kirill offline, seems like it would probably be worth just wontfixing this

**Comment by 333fred** [comment]
Alright. I'll close this out then.
